libraries/lib-registries/ClientData.h

```diff
@@ -410,6 +411,81 @@ class Site
       }
    }
 
+   //! Invoke function on corresponding pairs of ClientData objects that have
+   //! been created in @c this and in another Site with the same factories
+   /*!
+    Beware that the sequence of visitation is not specified.
+    @tparam Function takes two pointers to ClientData, return value is ignored
+    @param other supplies the objects to the second argument of function
+    @param function of type @b Function may assume the precondition, that the
+    arguments are not both null, and if create is true, then neither argument is
+    null.  When neither is null, also the objects have come from the same
+    factory.
+    @param create whether to create objects at vacant slots that correspond to
+    non-vacant slots.  (Never create where there are corresponding nulls.)
+    */
+   template<typename Function>
+   void ForCorresponding(Site &other, const Function &function,
+      bool create = true)
+   {
+      size_t size;
+      {
+         auto factories = GetFactories();
+         size = factories.mObject.size();
+         // Release lock on factories before getting one on data -- otherwise
+         // there would be a deadlock possibility inside EnsureIndex
+      }
+
+      // Lock two containers, carefully
+      std::optional<decltype(GetData())> oOtherData;
```

Comment:
I don't understand why comparing the addresses of two unrelated objects (both are Sites attached to a different objects, ain't they) is needed. I expect that this pointers are pretty much random and do not reflect any properties of the objects

Reply:
Site in general allows locking of accesses to slots depending on template arguments, although we do t use the feature anywhere yet.  This sequencing avoids possibility of deadlock.

Reply:
This deserves a very clear comment at least.